test/libgit2.jl

```diff
@@ -443,6 +443,60 @@ mktempdir() do dir
             end
         #end
 
+        #@testset "tree" begin
+            function addfile(root::Cstring, tentry::Ptr{Void}, payload::Ptr{Void})
+                strarr = unsafe_pointer_to_objref(payload)
+                sroot = unsafe_string(root)
+                te = LibGit2.GitTreeEntry(tentry)
+                fmode = LibGit2.filemode(te)
+                fname = LibGit2.filename(te)
+                if fmode == Cint(LibGit2.Consts.FILEMODE_BLOB)
+                    push!(strarr, fname)
+                end
+                return zero(Cint)
+            end
+
+            repo = LibGit2.GitRepo(test_repo)
+            head = LibGit2.head(repo)
+            ht = LibGit2.peel(LibGit2.GitTree, head)
+            try
+                # get tree entry by filename
+                @test isnull(LibGit2.lookup(ht, "no such name"))
+                tfte = get(LibGit2.lookup(ht, test_file))
+                try
+                    @test LibGit2.filename(tfte) == test_file
+                    @test LibGit2.filemode(tfte) == Cint(LibGit2.Consts.FILEMODE_BLOB)
+
+                    @test isnull(LibGit2.GitTreeEntry(ht, LibGit2.Oid()))
+                    tfoid = LibGit2.oid(tfte)
+                    tfte2 = LibGit2.GitTreeEntry(ht, tfoid)
+                    try
+                        @test !isnull(tfte2)
+                        @test LibGit2.filename(get(tfte2)) == test_file
+
+                        tfcontent = LibGit2.with(LibGit2.object(repo, get(tfte2))) do obj
+                            LibGit2.with(LibGit2.peel(LibGit2.GitBlob, obj)) do blob
+                                unsafe_string(convert(Cstring, LibGit2.content(blob)))
+                            end
+                        end
+                        @test startswith(tfcontent, commit_msg1)
+                    finally
+                        finalize(tfte2)
```

Comment:
Tree entries need to be freed before the tree object is freed.  But because entries created in try-block, a new try-block is created for finalizing them.  All of it can be rearranged so finalization can be postponed as much as possible.